ipaserver/install/ipa_idrange_fix.py

```diff
@@ -0,0 +1,1073 @@
+    print(horizontal_line)
+
+    # Draw the table rows
+    for id_range in id_ranges:
+        row = "| "
+        for column, width in max_widths.items():
+            value = getattr(id_range, column)
+            if value is not None:
+                row += f"{str(value).rjust(width)} | "
+            else:
+                # Adding the separator
+                row += " " * (width + 1) + "| "
+        print(row)
+    print(horizontal_line)
+# endregion
+# Reading from LDAP
+# region
+
+
+def read_ranges(suffix) -> list[IDRange]:
+    """Function to read ID ranges from LDAP"""
+    try:
+        ranges = api.Backend.ldap2.get_entries(
+            DN(api.env.container_ranges, suffix),
+            ldap.SCOPE_ONELEVEL,
+            "(objectclass=ipaIDRange)",
+        )
+    except errors.NotFound:
+        logger.error("LDAPError: No ranges found!")
+    except errors.ExecutionError as e:
+        logger.error("Exception while reading users: %s", e)
+    id_ranges: IDRange = []
+
+    for entry in ranges:
+        sv = entry.single_value
+        id_range = IDRange()
+        id_range.name = sv.get("cn")
+        id_range.size = int(sv.get("ipaidrangesize"))
+        id_range.first_id = int(sv.get("ipabaseid"))
+        id_range.base_rid = (
+            int(sv.get("ipabaserid")) if sv.get("ipabaserid") else None
+        )
+        id_range.secondary_base_rid = (
+            int(sv.get("ipasecondarybaserid"))
+            if sv.get("ipasecondarybaserid")
+            else None
+        )
+        id_range.suffix = suffix
+        id_range.type = sv.get("iparangetype")
+        id_range.dn = entry.dn
+
+        id_range.count()
+        logger.debug("ID range found: %s", id_range)
+
+        id_ranges.append(id_range)
+
+    id_ranges.sort(key=lambda x: x.first_id)
+    return id_ranges
+
+
+def read_outofrange_identities(suffix, id_ranges) -> list[IDentity]:
+    """Function to read out of range users and groups from LDAP"""
+    users_outofrange = read_ldap_ids(
+        DN(api.env.container_user, suffix),
+        True,
+        id_ranges
+    )
+    logger.info("Users out of range found: %s", len(users_outofrange))
+    del_outofrange = read_ldap_ids(
+        DN(api.env.container_deleteuser, suffix),
+        True,
+        id_ranges
+    )
+    logger.info("Preserved users out of range found: %s", len(del_outofrange))
+    groups_outofrange = read_ldap_ids(
+        DN(api.env.container_group, suffix),
+        False,
+        id_ranges
+    )
+    logger.info("Groups out of range found: %s", len(groups_outofrange))
+    outofrange = users_outofrange + del_outofrange + groups_outofrange
+    outofrange.sort(key=lambda x: x.number)
+    return outofrange
+
+
+def read_ldap_ids(container_dn, user: bool, id_ranges) -> list[IDentity]:
+    """Function to read IDs from containter in LDAP"""
+    id_entities = []
+    if user:
+        id_name = "user"
+        ldap_filter = get_outofrange_filter(id_ranges, "account", "uid")
```

Comment:
It would probably more flexible to use the full objectclass here and have get_outofrange_filter() do the combination instead of also adding "posix" to it. It is also more clear what the value does (I had to look).

Reply:
I think same for uidNumber/gidNumber.

Reply:
indeed, fixed.